src/sinks/appsignal/mod.rs

```diff
@@ -150,61 +163,279 @@ impl SinkConfig for AppsignalSinkConfig {
+    let request = Request::get(uri).header(AUTHORIZATION, format!("Bearer {}", push_api_key));
+    let response = client.send(request.body(Body::empty()).unwrap()).await?;
+
+    match response.status() {
+        status if status.is_success() => Ok(()),
+        other => Err(super::HealthcheckError::UnexpectedStatus { status: other }.into()),
+    }
 }
 
-impl HttpEventEncoder<serde_json::Value> for AppsignalEventEncoder {
-    fn encode_event(&mut self, mut event: Event) -> Option<serde_json::Value> {
-        self.transformer.transform(&mut event);
+struct AppsignalSink<S> {
+    service: S,
+    compression: Compression,
+    transformer: Transformer,
+    batch_settings: BatcherSettings,
+}
 
-        match event {
-            Event::Log(log) => Some(json!({ "log": log })),
-            Event::Metric(metric) => Some(json!({ "metric": metric })),
-            _ => panic!("The AppSignal sink does not support this type of event: {event:?}"),
-        }
+impl<S> AppsignalSink<S>
+where
+    S: Service<AppsignalRequest> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Response: DriverResponse + Send + 'static,
+    S::Error: std::fmt::Debug + Into<crate::Error> + Send,
+{
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let service = ServiceBuilder::new().service(self.service);
+
+        input
+            .batched(self.batch_settings.into_byte_size_config())
+            .request_builder(
+                None,
+                AppsignalRequestBuilder {
+                    compression: self.compression,
+                    encoder: AppsignalEncoder {
+                        transformer: self.transformer.clone(),
+                    },
+                },
+            )
+            .filter_map(|request| async move {
+                match request {
+                    Err(error) => {
+                        emit!(SinkRequestBuildError { error });
+                        None
+                    }
+                    Ok(req) => Some(req),
+                }
+            })
+            .into_driver(service)
+            .run()
+            .await
     }
 }
 
 #[async_trait::async_trait]
-impl HttpSink for AppsignalSinkConfig {
-    type Input = serde_json::Value;
-    type Output = Vec<BoxedRawValue>;
-    type Encoder = AppsignalEventEncoder;
-
-    fn build_encoder(&self) -> Self::Encoder {
-        AppsignalEventEncoder {
-            transformer: self.encoding.clone(),
+impl<S> StreamSink<Event> for AppsignalSink<S>
+where
+    S: Service<AppsignalRequest> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Response: DriverResponse + Send + 'static,
+    S::Error: std::fmt::Debug + Into<crate::Error> + Send,
+{
+    async fn run(
+        self: Box<Self>,
+        input: futures_util::stream::BoxStream<'_, Event>,
+    ) -> Result<(), ()> {
+        self.run_inner(input).await
+    }
+}
+
+#[derive(Clone)]
+struct AppsignalEncoder {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏ One thing we do with the new style refactors is break the logic into files/modules. You can see others as examples such as `kafka`.

Essentially:

`mod.rs`
`config.rs`
`encoder.rs`
`request_builder.rs`
`service.rs`
`sink.rs`
`tests.rs`
`integration_tests.rs`